File: msgpkids/code/SectorFive/SectorFive_4/enemy.rb
#---
# Excerpted from "Learn Game Programming with Ruby",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/msgpkids for more book information.
#---

class Enemy
  SPEED = 4
  attr_reader :x, :y, :radius
  def initialize(window)
    @y = 0
    @image = Gosu::Image.new(window,'images/enemy.png',false)
    @radius = 20
    @x = rand(window.width- 2 * @radius) + @radius
  end
  def draw
    @image.draw(@x - @radius, @y - @radius, 1)
  end
  def move
    @y += SPEED
  end
end

File: cdc-elixir/code/handle_the_uncertain/trycatch/dungeon_crawl/lib/dungeon_crawl/room/triggers/enemy.ex
#---
# Excerpted from "Learn Functional Programming with Elixir",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material,
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose.
# Visit http://www.pragmaticprogrammer.com/titles/cdc-elixir for more book information.
#---
defmodule DungeonCrawl.Room.Triggers.Enemy do
  @behaviour DungeonCrawl.Room.Trigger

  alias Mix.Shell.IO, as: Shell

  def run(character, %DungeonCrawl.Room.Action{id: :forward}) do
    enemy = Enum.random(DungeonCrawl.Enemies.all)

    Shell.info(enemy.description)
    Shell.info("The enemy #{enemy.name} wants to fight.")
    Shell.info("You were prepared and attack first.")
    {updated_char, _enemy} = DungeonCrawl.Battle.fight(character, enemy)

    {updated_char, :forward}
  end
end
